src/wire/ipv6option.rs

```diff
@@ -0,0 +1,356 @@
+        match packet.option_type() {
+            OptionType::Pad1 => {
+                Ok(Repr::Pad1)
+            }
+            OptionType::PadN => {
+                Ok(Repr::PadN {
+                    length: packet.option_data_length(),
+                    data: packet.data(),
+                })
+            }
+            _ => Err(Error::Unrecognized),
+        }
+    }
+
+    /// Return the length of a header that will be emitted from this high-level representation.
+    pub fn buffer_len(&self) -> usize {
+        match self {
+            &Repr::Pad1 => {
+                1
+            }
+            &Repr::PadN{length, ..} => {
+                (length as usize) + field::DATA.start
+            }
+
+            &Repr::__Nonexhaustive => unreachable!()
+        }
+    }
+
+    /// Emit a high-level representation into an IPv6 Extension Header Option packet.
+    pub fn emit<T: AsRef<[u8]> + AsMut<[u8]> + ?Sized>(&self, packet: &mut Packet<&'a mut T>) {
+        match self {
+            &Repr::Pad1 => {
+                packet.set_option_type(From::from(OptionType::Pad1));
+            }
+            &Repr::PadN{length, data} => {
+                let len = length as usize;
+                packet.set_option_type(From::from(OptionType::PadN));
+                packet.set_option_data_length(length);
+                packet.data_mut().copy_from_slice(&data[..len]);
+            }
+
+            &Repr::__Nonexhaustive => unreachable!()
+        }
+    }
+}
+
+impl<'a> fmt::Display for Repr<'a> {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        let name = "IPv6 Extension Option";
+        match self {
+            &Repr::Pad1 => {
+                write!(f, "{} type={} ", &name, OptionType::Pad1)
+            }
+            &Repr::PadN{length, ..} => {
+                write!(f, "{} type={} length={} ", &name, OptionType::PadN, length)
+            }
+
+            &Repr::__Nonexhaustive => unreachable!()
+        }
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+
+    static PACKET_BYTES_PAD1:    [u8; 1] = [0x0];
+    static PACKET_BYTES_PADN:    [u8; 3] = [0x1, 0x1, 0x0];
+    static PACKET_BYTES_UNKNOWN: [u8; 5] = [0xff, 0x3, 0x0, 0x0, 0x0];
+
+    #[test]
+    fn test_check_len() {
+        let bytes = [0u8];
+        // zero byte buffer
+        assert_eq!(Err(Error::Truncated), Packet::new(&bytes[..0]).check_len());
+        // pad1
+        assert_eq!(Ok(()), Packet::new(&PACKET_BYTES_PAD1).check_len());
+
+        // padn with truncated data
+        assert_eq!(Err(Error::Truncated), Packet::new(&PACKET_BYTES_PADN[..2]).check_len());
+        // padn
+        assert_eq!(Ok(()), Packet::new(&PACKET_BYTES_PADN).check_len());
+
+        // unknown option type with truncated data
+        assert_eq!(Err(Error::Truncated), Packet::new(&PACKET_BYTES_UNKNOWN[..4]).check_len());
+        assert_eq!(Err(Error::Truncated), Packet::new(&PACKET_BYTES_UNKNOWN[..1]).check_len());
```

Comment:
You could use `new_checked` here, but this works too.

Reply:
I decided to use `check_len` to ensure that it is working correctly independently of `new_checked`. I test `new_checked` in `test_option_deconstruct` too.

Reply:
Ah, makes sense